deepmd/dpmodel/model/make_model.py

```diff
@@ -475,11 +475,14 @@ def _format_nlist(
                 index = ret.reshape(n_nf, n_nloc * n_nnei, 1).repeat(3, axis=2)
                 coord1 = xp.take_along_axis(extended_coord, index, axis=1)
                 coord1 = coord1.reshape(n_nf, n_nloc, n_nnei, 3)
-                rr = xp.linalg.norm(coord0[:, :, None, :] - coord1, axis=-1)
-                rr = xp.where(m_real_nei, rr, float("inf"))
-                rr, ret_mapping = xp.sort(rr, axis=-1), xp.argsort(rr, axis=-1)
+                # jax raises NaN error using norm
+                # but note: we don't actually need to sqrt here; the squared value is enough
+                # rr = xp.linalg.norm(coord0[:, :, None, :] - coord1, axis=-1)
+                rr2 = xp.sum(xp.square(coord0[:, :, None, :] - coord1), axis=-1)
+                rr2 = xp.where(m_real_nei, rr2, float("inf"))
+                rr2, ret_mapping = xp.sort(rr2, axis=-1), xp.argsort(rr2, axis=-1)
                 ret = xp.take_along_axis(ret, ret_mapping, axis=2)
-                ret = xp.where(rr > rcut, -1, ret)
+                ret = xp.where(rr2 > rcut * rcut, -1, ret)
```

Comment:
I have to modify here to fix NaN forces. However, I don't quite understand why the energy has a gradient to the neighbor llist. Did I miss something?

Reply:
Anyway, I think avoiding sqrt also improves performance.

Reply:
Interesting. I agree with you.

Reply:
stop gradient at the nlist? https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html

Reply:
I found I couldn't reproduce the original error after reverting this piece of code. Not sure the reason...

Reply:
But I think the reason of NaN is that the neighbor list is rebuilt in the linear atomic model.

Reply:
Ok, 84803b847e641cf5a688bf2b6b9c16f3632b25c1 should fix it. I am wondering whether nlist in pytorch also has the gradient.